modules/build/src/main/scala/scala/build/input/Inputs.scala

```diff
@@ -386,15 +347,70 @@ object Inputs {
         case Right(elem) => elem
       }.flatten
       assert(extraClasspathWasPassed || validElems.nonEmpty)
+      val (inferredWorkspace, inferredNeedsHash, workspaceOrigin) = {
+        val settingsFiles = validElems.projectSettingsFiles
+        val dirsAndFiles = validElems.collect {
+          case d: Directory  => d
+          case f: SourceFile => f
+        }
+        settingsFiles.headOption.map { s =>
+          if (settingsFiles.length > 1)
+            System.err.println(
+              s"Warning: more than one ${Constants.projectFileName} file has been found. Setting ${s.base} as the project root directory for this run."
+            )
+          (s.base, true, WorkspaceOrigin.SourcePaths)
+        }.orElse {
+          dirsAndFiles.collectFirst {
+            case d: Directory =>
+              if (dirsAndFiles.length > 1)
+                System.err.println(
+                  s"Warning: setting ${d.path} as the project root directory for this run."
+                )
+              (d.path, true, WorkspaceOrigin.SourcePaths)
+            case f: SourceFile =>
+              if (dirsAndFiles.length > 1)
+                System.err.println(
+                  s"Warning: setting ${f.path / os.up} as the project root directory for this run."
+                )
+              (f.path / os.up, true, WorkspaceOrigin.SourcePaths)
+          }
+        }.orElse {
+          validElems.collectFirst {
+            case _: Virtual =>
+              (os.temp.dir(), true, WorkspaceOrigin.VirtualForced)
+          }
+        }.getOrElse((os.pwd, true, WorkspaceOrigin.Forced))
+      }
 
-      Right(forValidatedElems(
-        validElems,
-        baseProjectName,
-        forcedWorkspace,
-        enableMarkdown,
-        allowRestrictedFeatures,
-        extraClasspathWasPassed
-      ))
+      val (workspace, needsHash, workspaceOrigin0) = forcedWorkspace match {
+        case None => (inferredWorkspace, inferredNeedsHash, workspaceOrigin)
+        case Some(forcedWorkspace0) =>
+          val needsHash0 = forcedWorkspace0 != inferredWorkspace || inferredNeedsHash
+          (forcedWorkspace0, needsHash0, WorkspaceOrigin.Forced)
+      }
+
+      if !Properties.isWin && workspace.toString.contains(":") then
```

Comment:
Do we allow it on windows? I think in that case classpath separator is `;` , so maybe we should have a different one per system?

Reply:
fixed, should account for the Windows class path separator now